moveit_ros/moveit_servo/src/utilities.cpp

```diff
@@ -179,4 +179,82 @@ bool applyJointUpdate(rclcpp::Clock& clock, const double publish_period, const E
+                             const Eigen::Isometry3d& tf_moveit_to_ee_frame,
+                             const moveit::core::RobotStatePtr& current_state)
+{
+  Eigen::Vector3d translation_vector(cmd.twist.linear.x, cmd.twist.linear.y, cmd.twist.linear.z);
+  Eigen::Vector3d angular_vector(cmd.twist.angular.x, cmd.twist.angular.y, cmd.twist.angular.z);
+
+  // If the incoming frame is empty or is the command frame, we use the previously calculated tf
+  if (cmd.header.frame_id.empty() || cmd.header.frame_id == servo_params.robot_link_command_frame)
+  {
+    translation_vector = tf_moveit_to_robot_cmd_frame.linear() * translation_vector;
+    angular_vector = tf_moveit_to_robot_cmd_frame.linear() * angular_vector;
+  }
+  else if (cmd.header.frame_id == servo_params.ee_frame_name)
+  {
+    // If the frame is the EE frame, we already have that transform as well
+    translation_vector = tf_moveit_to_ee_frame.linear() * translation_vector;
+    angular_vector = tf_moveit_to_ee_frame.linear() * angular_vector;
+  }
+  else
+  {
+    // We solve (planning_frame -> base -> cmd.header.frame_id)
+    // by computing (base->planning_frame)^-1 * (base->cmd.header.frame_id)
+    const auto tf_moveit_to_incoming_cmd_frame =
+        current_state->getGlobalLinkTransform(servo_params.planning_frame).inverse() *
+        current_state->getGlobalLinkTransform(cmd.header.frame_id);
+    translation_vector = tf_moveit_to_incoming_cmd_frame.linear() * translation_vector;
+    angular_vector = tf_moveit_to_incoming_cmd_frame.linear() * angular_vector;
+  }
+
+  // Put these components back into a TwistStamped
+  cmd.header.frame_id = servo_params.planning_frame;
+  cmd.twist.linear.x = translation_vector(0);
+  cmd.twist.linear.y = translation_vector(1);
+  cmd.twist.linear.z = translation_vector(2);
+  cmd.twist.angular.x = angular_vector(0);
+  cmd.twist.angular.y = angular_vector(1);
+  cmd.twist.angular.z = angular_vector(2);
+}
+
+geometry_msgs::msg::Pose deltaToPose(const Eigen::VectorXd& delta_x, const Eigen::Isometry3d& ik_base_to_tip_frame)
+{
+  // get a transformation matrix with the desired position change &
+  // get a transformation matrix with desired orientation change
+  Eigen::Isometry3d tf_pos_delta(Eigen::Isometry3d::Identity());
+  tf_pos_delta.translate(Eigen::Vector3d(delta_x[0], delta_x[1], delta_x[2]));
+
+  Eigen::Isometry3d tf_rot_delta(Eigen::Isometry3d::Identity());
+  Eigen::Quaterniond q = Eigen::AngleAxisd(delta_x[3], Eigen::Vector3d::UnitX()) *
+                         Eigen::AngleAxisd(delta_x[4], Eigen::Vector3d::UnitY()) *
+                         Eigen::AngleAxisd(delta_x[5], Eigen::Vector3d::UnitZ());
+  tf_rot_delta.rotate(q);
+
+  // Poses passed to IK solvers are assumed to be in some tip link (usually EE) reference frame
+  // First, find the new tip link position without newly applied rotation
+  auto tf_no_new_rot = tf_pos_delta * ik_base_to_tip_frame;
```

Comment:
Can you please use full types instead of `auto` here? It makes the code more readable. Easier to know if these things are vectors or transforms.

```suggestion
  Eigen::Isometry3d tf_no_new_rot = tf_pos_delta * ik_base_to_tip_frame;
```